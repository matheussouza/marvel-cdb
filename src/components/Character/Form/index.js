import React, { useState } from 'react';

export default ({ character, back, update }) => {
  const [name, setName] = useState(character.name);
  const [description, setDescription] = useState(character.description || '');

  const confirmCancel = () => {
    return window.confirm('Your changes will be lost. Do you want to proceed?');
  }

  return (
    <div>
      <div className="d-flex align-items-center justify-content-between">
        <div>
          <h3 className="m-0 d-inline">Update data</h3>
        </div>
        <div className="d-none d-sm-block">
          <button
            data-testid="save-computer"
            className="btn btn-primary"
            onClick={ () => {
              update(name, description);
              back();
            }}
          >
            Save
          </button>
          <button
            data-testid="cancel-computer"
            className="btn btn-secondary"
            onClick={() => {
              if (confirmCancel())
                back();
            }}
            style={{ marginLeft: 10 }}
          >
            Cancel
          </button>
        </div>
      </div>
      <hr />
      <div>
        <div className="form-group">
          <label>Name</label>
          <input data-testid="name-input" type="text" className="form-control" value={name} onChange={ (event) => setName(event.target.value) } />
        </div>

        <div className="form-group">
          <label>Description</label>
          <textarea
            data-testid="description-input"
            className="form-control"
            onChange={ (event) => setDescription(event.target.value) }
            value={description || 'Description unavailable.'}
          />
        </div>
        <div className="d-sm-none" style={{ paddingBottom: 20 }}>
          <button
            data-testid="save-mobile"
            className="btn btn-primary"
            onClick={ () => {
              update(name, description);
              back();
            }}
          >
            Save
          </button>
          <button
            data-testid="cancel-mobile"
            className="btn btn-secondary"
            onClick={() => {
              if (confirmCancel())
                back();
            }}
            style={{ marginLeft: 10 }}
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  )
}
