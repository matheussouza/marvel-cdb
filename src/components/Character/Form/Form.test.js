import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import Form from './index'

describe('Character Form component', () => {

  let backMock;
  let updateMock;

  beforeEach(() => {
    backMock = jest.fn();
    updateMock = jest.fn();
    global.confirm = jest.fn();
  });

  it('should exist', () => {
    const subject = render(<Form character={{}} />);
    expect(subject.container).toMatchSnapshot();
  });

  it('should set input with name', () => {
    const subject = render(<Form character={{ name: 'Spider-Man' }} />);
    expect(subject.getByTestId('name-input')).toHaveProperty('value', 'Spider-Man');
  });

  it('should set input with description', () => {
    const subject = render(<Form character={{ description: 'An optional description' }} />);
    expect(subject.getByTestId('description-input')).toHaveProperty('value', 'An optional description');
  });
  
  it('should have default description', () => {
    const subject = render(<Form character={{}} />);
    expect(subject.getByTestId('description-input')).toHaveProperty('value', 'Description unavailable.');
  });

  it('should update on click (computer)', () => {
    const subject = render(<Form character={{ name: 'Spider-Man', description: 'My Description.' }} update={ updateMock } back={ backMock } />);
    
    expect(backMock).toBeCalledTimes(0);
    expect(updateMock).toBeCalledTimes(0);
    fireEvent.click(subject.getByTestId('save-computer'));
    expect(backMock).toBeCalledTimes(1);
    expect(updateMock).toBeCalledTimes(1);
    expect(updateMock).toBeCalledWith('Spider-Man', 'My Description.');
  });

  it('should update on click (mobile)', () => {
    const subject = render(<Form character={{ name: 'Spider-Man', description: 'My Description.' }} update={ updateMock } back={ backMock } />);
    
    expect(backMock).toBeCalledTimes(0);
    expect(updateMock).toBeCalledTimes(0);
    fireEvent.click(subject.getByTestId('save-mobile'));
    expect(backMock).toBeCalledTimes(1);
    expect(updateMock).toBeCalledTimes(1);
    expect(updateMock).toBeCalledWith('Spider-Man', 'My Description.');
  });

  it('should cancel on click (computer)', () => {
    const subject = render(<Form character={{}} back={ backMock } />);
    global.confirm = jest.fn(() => true);
    
    expect(global.confirm).toBeCalledTimes(0);
    expect(backMock).toBeCalledTimes(0);
    fireEvent.click(subject.getByTestId('cancel-computer'));
    expect(global.confirm).toBeCalledTimes(1);
    expect(backMock).toBeCalledTimes(1);
  });

  it('should cancel on click (mobile)', () => {
    const subject = render(<Form character={{}} back={ backMock } />);
    global.confirm = jest.fn(() => true);
    
    expect(global.confirm).toBeCalledTimes(0);
    expect(backMock).toBeCalledTimes(0);
    fireEvent.click(subject.getByTestId('cancel-mobile'));
    expect(global.confirm).toBeCalledTimes(1);
    expect(backMock).toBeCalledTimes(1);
  });

  it('should\'t cancel on click (computer) - select not', () => {
    const subject = render(<Form character={{}} back={ backMock } />);
    global.confirm = jest.fn(() => false);
    
    expect(global.confirm).toBeCalledTimes(0);
    expect(backMock).toBeCalledTimes(0);
    fireEvent.click(subject.getByTestId('cancel-computer'));
    expect(global.confirm).toBeCalledTimes(1);
    expect(backMock).toBeCalledTimes(0);
  });

  it('should\'t cancel on click (mobile) - select not', () => {
    const subject = render(<Form character={{}} back={ backMock } />);
    global.confirm = jest.fn(() => false);
    
    expect(global.confirm).toBeCalledTimes(0);
    expect(backMock).toBeCalledTimes(0);
    fireEvent.click(subject.getByTestId('cancel-mobile'));
    expect(global.confirm).toBeCalledTimes(1);
    expect(backMock).toBeCalledTimes(0);
  });

});