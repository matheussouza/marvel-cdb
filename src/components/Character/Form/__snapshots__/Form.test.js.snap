// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Form component should exist 1`] = `
<div>
  <div>
    <div
      class="d-flex align-items-center justify-content-between"
    >
      <div>
        <h3
          class="m-0 d-inline"
        >
          Update data
        </h3>
      </div>
      <div
        class="d-none d-sm-block"
      >
        <button
          class="btn btn-primary"
          data-testid="save-computer"
        >
          Save
        </button>
        <button
          class="btn btn-secondary"
          data-testid="cancel-computer"
          style="margin-left: 10px;"
        >
          Cancel
        </button>
      </div>
    </div>
    <hr />
    <div>
      <div
        class="form-group"
      >
        <label>
          Name
        </label>
        <input
          class="form-control"
          data-testid="name-input"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label>
          Description
        </label>
        <textarea
          class="form-control"
          data-testid="description-input"
        >
          Description unavailable.
        </textarea>
      </div>
      <div
        class="d-sm-none"
        style="padding-bottom: 20px;"
      >
        <button
          class="btn btn-primary"
          data-testid="save-mobile"
        >
          Save
        </button>
        <button
          class="btn btn-secondary"
          data-testid="cancel-mobile"
          style="margin-left: 10px;"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
